okhttp/src/main/java/com/squareup/okhttp/IdnaMappingTable.java

```diff
@@ -0,0 +1,394 @@
+ * a 16 bit range. For example, values in 0x30000..0x3ffff are in page 0x3 which is then binary
+ * searched over a 16 bit suffix: between 0x0000..0xffff. We save ~16 KiB of memory using unsigned
+ * shorts instead of ints as the subjects of our binary search.
+ *
+ * <p>The value of each range is also an unsigned short, containing a status (valid, ignored,
+ * mapped, etc.) and either an IDNA2008 status (NV8 or XV8) or an offset to look up the replacement
+ * string. All replacement strings are concatenated substrings of a shared ~8,600 character string.
+ *
+ * <p>This class uses bit packing to store several values in a single unsigned short. There are two
+ * bit packing choices:
+ * <ul>
+ *   <li>Ranges with status "mapped" or "disallowed_STD3_mapped" use 2 bits for the bit packing type
+ *       and status, plus 14 bits for the offset of the mapped string.
+ *   <li>Other ranges use 2 bits for the bit packing type, 6 bits that are unused, 4 bits for IDNA
+ *       2008 status and 4 bits for status.
+ * </ul>
+ *
+ * <p>This documentation refers to "unsigned short". Java's unsigned short type is named 'char',
+ * but none of the arrays in this class contain UTF-16 code points and should not be interpreted
+ * as text.
+ */
+final class IdnaMappingTable {
+  private static final BitPackingType[] BIT_PACKING_TYPE_VALUES = BitPackingType.values();
+  private static final Status[] STATUS_VALUES = Status.values();
+  private static final Idna2008Status[] IDNA_2008_STATUS_VALUES = Idna2008Status.values();
+  public static final IdnaMappingTable INSTANCE;
+  static {
+    try {
+      BufferedSource source = Okio.buffer(Okio.source(
+          IdnaMappingTable.class.getResourceAsStream("/IdnaMappingTable.bin")));
+      INSTANCE = IdnaMappingTable.readBinary(source);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to initialize IDNA mapping table.", e);
+    }
+  }
+
+  private final char[] pageOffsets;
+  private final char[] rangeStarts;
+  private final char[] rangeValues;
+  private final String replacements;
+
+  private IdnaMappingTable(
+      char[] pageOffsets, char[] rangeStarts, char[] rangeValues, String replacements) {
+    this.pageOffsets = pageOffsets;
+    this.rangeStarts = rangeStarts;
+    this.rangeValues = rangeValues;
+    this.replacements = replacements;
+  }
+
+  /** Write this table to {@code out}. */
+  public void writeBinary(BufferedSink out) throws IOException {
+    out.writeInt(pageOffsets.length);
+    for (char c : pageOffsets) {
+      out.writeShort(c);
+    }
+
+    out.writeInt(rangeStarts.length);
+    for (char c : rangeStarts) {
+      out.writeShort(c);
+    }
+    for (char c : rangeValues) {
+      out.writeShort(c);
+    }
+
+    Buffer replacementsUtf8 = new Buffer().writeUtf8(replacements);
+    out.writeInt((int) replacementsUtf8.size());
+    out.write(replacementsUtf8, replacementsUtf8.size());
```

Comment:
Yes! Think there's much utility to bulk APIs for arrays? To potentially speed up `out.writeShort()` calls in the above loops.